pkg/operator/staticpod/startupmonitor/startup_monitor.go

```diff
@@ -0,0 +1,370 @@
+// allows for initialization and an operand's health condition assessment
+type HealthChecker interface {
+	// Start is used to initialize and start the underlying health monitor
+	Start(context.Context) error
+
+	// IsTargetHealthy defines a function that abstracts away assessing operand's health condition.
+	// the provided functions should be async and cheap in a sense that it shouldn't assess the target
+	// only read the current state.
+	// mainly because we acquire a lock on each sync.
+	IsTargetHealthy() (healthy bool, reason string, err error)
+}
+
+// WantsRestConfig an optional interface used for setting rest config for Kube API
+type WantsRestConfig interface {
+	SetRestConfig(config *rest.Config)
+}
+
+// monitor is a controller that watches an operand's health condition
+// and falls back to the previous version in case the current version is considered unhealthy.
+//
+// This controller understands a tree structure created by an OCP installation. That is:
+//  The root manifest are looked up in the manifestPath
+//  The revisioned manifest are looked up in the staticPodResourcesPath
+//  The target (operand) name is derived from the targetName.
+type monitor struct {
+	// probeInterval specifies a time interval at which health of the target will be assessed
+	// be mindful of not setting it too low, on each iteration, an i/o is involved
+	probeInterval time.Duration
+
+	// timeout specifies a timeout after which the monitor starts the fall back procedure
+	timeout time.Duration
+
+	// revision at which the monitor was started
+	revision int
+
+	// targetName hold the name of the operand
+	// used to construct the final file name when reading the current and previous manifests
+	targetName string
+
+	// manifestsPath points to the directory that holds the root manifests
+	manifestsPath string
+
+	// staticPodResourcesPath points to the directory that holds all files supporting the static pod manifests
+	staticPodResourcesPath string
+
+	// healthChecker is used to get the current operand's health condition
+	healthChecker HealthChecker
+
+	// records the time the monitor has started assessing operand's health condition
+	monitorTimeStamp time.Time
+
+	// io collects file system level operations that need to be mocked out during tests
+	io ioInterface
+}
+
+func newMonitor(restConfig *rest.Config, healthChecker HealthChecker) *monitor {
+	if wants, ok := healthChecker.(WantsRestConfig); ok {
+		wants.SetRestConfig(restConfig)
+	}
+
+	return &monitor{healthChecker: healthChecker, io: realFS{}}
+}
+
+func (m *monitor) Run(ctx context.Context) error {
+	klog.Infof("Starting the startup monitor with Interval = %v, Timeout = %v", m.probeInterval, m.timeout)
+	defer klog.Info("Shutting down the startup monitor")
+
+	if err := m.healthChecker.Start(ctx); err != nil {
+		return fmt.Errorf("failed to start the health checker for %s, due to %v", m.targetName, err)
+	}
+
+	wait.Until(m.syncErrorWrapper, m.probeInterval, ctx.Done())
```

Comment:
wait.Until has a fixed interval for the start of the next round, i.e. if sync takes longer than probeInterval, this will hotloop. There is a wait variant that has a constant delay between sync end and next sync.

Reply:
And shouldn't this be `wait.PollSomeThing` with a bool return value?